[AutoDiff] Fix `wrt` differentiation parameter clause printing.

dan-zheng opened the pull request:
Fix `wrt` parameter clause printing for `@differentiable` attribute.
- Previously, only parsed parameters (`ArrayRef<ParsedAutoDiffParameter>`) were used.
- Now, `AutoDiffParameterIndices` are used and preferred, while maintaining the same
  printing behavior.
  - Always print `@differentiable`; never `@differentiable()`.
  - Always print `@differentiable(wrt: x)`; never `@differentiable(wrt: (x))`.

Add tests, update affected existing tests.

Up next: fix `wrt` printing/serialization for `@differentiating` attribute.

dan-zheng committed:
[AutoDiff] Fix `wrt` differentiation parameter clause printing.

Fix `wrt` parameter clause printing for `@differentiable` attribute.
- Previously, only parsed parameters (`ArrayRef<ParsedAutoDiffParameter>`) were used.
- Now, `AutoDiffParameterIndices` are used and preferred, while maintaining the same
  printing behavior.
  - Always print `@differentiable`; never `@differentiable()`.
  - Always print `@differentiable(wrt: x)`; never `@differentiable(wrt: (x))`.

Add tests, update affected existing tests.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on test/AutoDiff/differentiable_attr_type_checking.swift:
```diff
@@ -590,12 +590,12 @@ protocol DiffReq : Differentiable {
 // expected-error @+1 {{does not conform to protocol 'DiffReq'}}
 struct ConformingWithErrors : DiffReq {
-  // expected-note @+1 {{candidate is missing attribute '@differentiable(wrt: (self, x))'}}
+  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
```
The PR description says it's maintaining the same printing behavior, but it seems not to be the case. We used `ParsedParameterIndices` because diagnostics should be WYSIWIG from the protocol declaration. Is there a way to still maintain that behavior?

dan-zheng replied:
We discussed in person.

We agreed that the new printing behavior is desirable, because the printed `@differentiable` is shorter. However, it's necessary to account for cross-module differences in parameter type differentiability; it should be valid to directly use the printed `@differentiable` attribute.

Done in https://github.com/apple/swift/pull/23379/commits/c989899bdbe0c073500ff407a56bbe9d468702da. Cross-module test added.

dan-zheng committed:
[AutoDiff] Enable pretty-printing `@differentiable` in a specific module.

Enable pretty-printing `@differentiable` attribute in a specific module.

This is important to ensure that diagnostics show `@differentiable` attributes
with accurate `wrt` parameters.

Add cross-module `@differentiable` protocol requirement test.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
